### DISCUSSION TITLE
Comparison with duchess

### DISCUSSION THREAD

**Discussion by nikomatsakis** [giovanniberti/robusta#68] [General]
Hi @giovanniberti and others! I've been working on this project that has very similar goals:

https://github.com/duchess-rs/duchess

I'd love to talk to you about how we can join forces! Maybe we can schedule a time to chat live? (You can reach me at `rust@nikomatsakis.com`, tbh I'm struggling to keep up with GH notifications right now, but I will try!)

**Comment by nikomatsakis** [comment]
Oh, and I will be at RustNation, so maybe the easiest thing to do is just to sync there!

**Comment by uvlad7** [comment]
It looks interesting. I'm currently working #50 on robusta, but only because I need something convenient to create a lot of Rust bindings to Java classes. I want to make it [possible](https://github.com/uvlad7/xml2json-rb/tree/feature/jruby) to implement Jruby methods in Rust, and while I decided to use my own approach to implement `native` methods - `env.register_native_methods` with some macroses, - I need something to call Java methods, wrap Java classes and so on, since Jruby has only Java API.

So, here are some pitfalls I have found:

- Arrays of wrapped objects, probably multidimensional
- Null values
- Generis #51, it's not possible to wrap ArrayList<MyType> and have proper types on the Rust side.
- Overloaded methods #72
- Possibility to throw exceptions with dynamic class and message #30 
- Signature validation for native methods, `System.loadLibrary` should fail, currently only name is validated, there is a possibility to define a long name, though, see #72

The last two are why I have chosen my own solution for native methods. Well, not only because of that, but also because it's closer to how it's implemented in [magnus](https://github.com/matsadler/magnus), binding for C Ruby.

Fields can be a problem too, I've seen a [note](https://github.com/giovanniberti/robusta/blob/fc5007c553bcb27dfb8673fbcad32febde583533/src/convert/field.rs#L78) in the source code, wasn't a problem for me yet, though. And probably inheritance #74.

So, I'm curious if duchess is better for my needs. I see it has some codegen magic, will it work without java source code? Jruby only have `jruby.jar` in its installation. Chaining java methods calls looks promising, for example, that's a feature I was thinking about. But is it gonna be convenient for wrapping java code that's not designed for that? Actually, I want interop code to be simple enough to be able to create some sort of `jbindgen` and generate it (almost) automatically.
